src/Datadog.Trace.ClrProfiler.Managed/AutoInstrumentation/AdoNet/AdoNetClientInstrumentMethodAttribute.cs

```diff
@@ -0,0 +1,125 @@
+using System;
+
+namespace Datadog.Trace.ClrProfiler.AutoInstrumentation.AdoNet
+{
+    internal class AdoNetClientInstrumentMethodAttribute : InstrumentMethodAttribute
+    {
+        public AdoNetClientInstrumentMethodAttribute(Type adoNetClientDataType)
+        {
+            AdoNetClientData = (IAdoNetClientData)Activator.CreateInstance(adoNetClientDataType);
+
+            AssemblyName = AdoNetClientData.AssemblyName;
+            TypeName = AdoNetClientData.SqlCommandType;
+            MinimumVersion = AdoNetClientData.MinimumVersion;
+            MaximumVersion = AdoNetClientData.MaximumVersion;
+            IntegrationName = AdoNetClientData.IntegrationName;
+        }
+
+        protected IAdoNetClientData AdoNetClientData { get; }
+
+        internal class CommandExecuteNonQueryAsyncAttribute : AdoNetClientInstrumentMethodAttribute
+        {
+            public CommandExecuteNonQueryAsyncAttribute(Type adoNetClientDataType)
+                : base(adoNetClientDataType)
+            {
+                MethodName = AdoNetConstants.MethodNames.ExecuteNonQueryAsync;
```

Comment:
Instead of assigning these values in constructors, could the public properties just be overridden? This infrastructure seems like more work than is needed, but perhaps I don't have the full understanding

Reply:
Oh wait, you're trying to re-use one attribute to represent the same method but just have different assembly names and return types. Okay so we do need some sort of parameterization to make it simple like you're envisioning

Reply:
well this is only done once for all possible ADO.NET integrations, the usage of these attributes are in the SqlClientDefinitions.cs file.

Reply:
Got it. So I guess my only feedback here is, would it make sense to override the properties for the constants that do not depend on the `adoNetClientDataType` and then in the constructor set the ones that do? It's a small refactor that might might improve readability, in my opinion

Reply:
Oh sure, I'll do that

Reply:
do you prefer it like this?

![image](https://user-images.githubusercontent.com/69803/106204608-8996b280-61bd-11eb-91f6-7dbf828289ab.png)

Reply:
done in 15c7af7

Reply:
Sorry, I didn't see this in time! After seeing the effects of the changes, I don't think it improves the readability like I thought it did. If you prefer this way let's keep it, otherwise my slight preference is to revert this latest commit. Sorry to waste your time like that

Reply:
don't worry that commit has been reverted. ;)